### PR TITLE
 V2.0.0

### PR DESCRIPTION
**Pull request by cfnelson**
**NOTE: WIP**

**Changes Made:**
- Remove embedded `analytics.js`
- Refactor `@okgrow/auto-analytics` to expect `analytics.js` to be passed in.
- CHANGELOG.md added
- Readme updated for v2.0.0 with instructions on how to build your own `analytics.min.js`
- Examples updated to work for v2.0.0 (tested locally by removing the `@okgrow/auto-analytics` deps and using the `dist/index.js` file)
- Removed `mocha, chai, nodemon, mock-browser` and replaced with `jest`.
- Install, setup and implement tests with`Jest`. 🌮 🎉
- Investigated & fixed existing bugs & buggy behaviour when react-helmet or any changes to `document.title` occur after route change. Will port fix back to make a patch release.
- Referrer now works correctly if browser's bck/fwd buttons are used. Will port fix back to make a patch release.

**TODO:**

- [x] ~Merge this [PR #1](https://github.com/okgrow/analytics.js/pull/1) first, as we reference it in our instructions.~ 🔥 
- [x] Review & discuss current design.
- [ ] Discuss testing strategy & what further tests should be created.
- [x] Have others follow the ReadMe instructions and see if they can get analytics to work.
- [x] Update Changelog & pkg to include the new version.

### CODE REVIEW

**Comment by RichardLitt** [comment]
If we're going to run with a changelog, perhaps we should implement [Semantic Release](https://github.com/semantic-release/semantic-release) to make it easier to build automatically?

File: src/index.js [hunk 1]
```diff
@@ -111,16 +126,23 @@ const bootstrapAnalytics = () => {
...
 // Make our helpers available
 export { trackEventWhenReady, trackPageWhenReady, identifyWhenReady };
 
-export default function (settings) {
+export default function ({ analytics, settings }) {
+  // TODO: Improve detection of incorrect params & provide warnings.
+  if (typeof analytics !== 'object' || typeof settings !== 'object') {
+    console.error('Analytics is not logging! You must initialize your analytics correctly.');
+    return;
+  }
```
Comment (ghost) [inline src/index.js]: Check for the expected functions we need.

**Comment by cfnelson** [comment]
@RichardLitt Thanks for the suggestion, I have asked everyone on the team to take a look and provide their thoughts. For the time being I am happy to manually maintain a changelog. I do want to end up with a automated process, just need to find the optimal solution that works for everyone.

File: .eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+  "extends": "airbnb-base",
+  "rules": {
+    "max-len": 1, // Sometimes necessary to have long strings and not risk whitespace
+    "no-restricted-globals": 1, // Allow usage of globals like location
```
Comment (ghost) [inline .eslintrc.js:5]: Smart move.

File: README.md [hunk 3]
```diff
@@ -22,66 +22,84 @@ Use one API, thanks to Segment.io's [analytics.js](https://segment.com/docs/libr
...
 
-## Ad-blocker
+OKGAnalytics(settings);
+```
 
-If you, or your users, are running an ad blocker in their browser and the analytics package is not bundled into a single JavaScript file to the browser (i.e., is downloads as analytics.js or something similar) the browser's ad blocker may prevent analytics tracking. This can happen during development mode when all JavaScript files are typically not bundled together.
+## Creating Segment's analytics.js
 
-To solve this problem with a Meteor application, for example, you can run the application in production mode like this:
+### Recommended process
```
Comment (ghost) [inline README.md:65]: Nice addition!

File: README.md [hunk 4]
```diff
@@ -22,66 +22,84 @@ Use one API, thanks to Segment.io's [analytics.js](https://segment.com/docs/libr
...
+Too reduce the final bundle size, remove any integrations that you are not using. To do that:
 
-## Configuration
+1. Open up the `lib/integrations.js` file and remove all the integrations that you will not use.
+1. Now re-run `npm run build`
+1. Your `analytics.js` and `analytics.min.js` will only contain the integrations you are using.
+1. Copy the `analytics.min.js` file to your project so that `@okgrow/auto-analytics` can use it.
 
-Add various platforms by adding each tool's configuration to the settings object passed to OKGAnalytics:
+### Not Recommended: Use Segment's example `analytics.js` Package
```
Comment (ghost) [inline README.md:79]: Agreed based on our previous discussion.

File: README.md [hunk 5]
```diff
@@ -22,66 +22,84 @@ Use one API, thanks to Segment.io's [analytics.js](https://segment.com/docs/libr
...
+  analytics,
+  integrations: {
+    'Google Analytics': { 'trackingId': 'Your tracking ID' },
+    'Mixpanel': { 'token': '...', 'people': true },
+  },
+  options: {
+    // Segment options to be passed to initialize() from analytics.js-core
+  },
+  autorun: true, // Defaults to true if not provided.
+};
```
Comment (ghost) [inline README.md:58]: This looks like a slightly different structure from what I see on CHANGELOG.md:10. There I see an object pass which will have two keys/properties: `analytics` (the analytics object) and `settings` (which I presume would _contain_ the three other keys/properties I see here: `integrations`, `options` and `autorun`.)

File: examples/react-router/package.json [hunk 6]
```diff
@@ -10,8 +10,15 @@
     "meteor-node-stubs": "~0.2.0",
     "react": "^15.4.2",
     "react-dom": "^15.4.2",
-    "react-komposer": "^2.0.0",
     "react-router": "^3.0.2",
-    "@okgrow/auto-analytics": "^1.0.2"
+    "@okgrow/auto-analytics": "^1.0.0"
```
Comment (ghost) [inline examples/react-router/package.json]: Should this be 2.0.0?

File: examples/react-router/client/main.jsx [hunk 7]
```diff
@@ -6,17 +6,26 @@ import { Random } from 'meteor/random';
...
-const SETTINGS = (Meteor.settings
+const integrations = (Meteor.settings
                   && Meteor.settings.public
                   && Meteor.settings.public.analyticsSettings)
                   || false;
 
-// Initialize the @okgrow/auto-analytics package
-OKGAnalytics(SETTINGS);
+// NOTE: Initialize the @okgrow/auto-analytics package.
+OKGAnalytics({ analytics, integrations, options: {}, autorun: true });
```
Comment (ghost) [inline examples/react-router/client/main.jsx]: So it looks like the `CHANGELOG.md` needs to be updated.

File: src/__tests__/index.js [hunk 8]
```diff
@@ -1,63 +1,86 @@
...
+  autorun: true,
+};
 
 // Some very basic "smoke test" stuff...
-describe('okgrow-auto-analytics imports', () => {
-  it('analytics is an object', async () => {
-    OKGAnalytics.analytics.should.be.a('object');
+describe('@okgrow/auto-analytics correctly exposes', () => {
+  test('trackEventWhenReady is a function', () => {
+    expect(trackEventWhenReady).toBeDefined();
```
Comment (ghost) [inline src/__tests__/index.js:22]: I suppose we _could_ add something like `expect(typeof trackEventWhenReady).toBe('function')` But that might be overkill.

File: src/__tests__/index.js [hunk 9]
```diff
@@ -1,63 +1,86 @@
...
+  const page = jest.fn();
+  const ready = jest.fn();
+  const track = jest.fn();
+  const identify = jest.fn();
+
+  // NOTE: Only unhandled edge case is if a user is to pass null.
+  // As a TypeError will be thrown due to destructuring.
+  test('Throws error when no params passed', () => {
+    const init = () => OKGAnalytics();
+    expect(init).toThrowError('Analytics is not logging! You must initialize analytics with the correct params.');
```
Comment (ghost) [inline src/__tests__/index.js]: Nice!

File: src/__tests__/index.js [hunk 10]
```diff
@@ -1,63 +1,86 @@
...
+    expect(missingPage).toThrowError('Analytics is not logging! Expected analytics to contain page , function(s).');
+  });
 
-    // eslint-disable-next-line no-undef
-    window.onload.should.be.a('function');
+  test('Throws when analytics is missing identify()', () => {
+    const missingIdentify = () => OKGAnalytics({ analytics: { ready, track, page }, integrations: {} });
+    expect(missingIdentify).toThrowError('Analytics is not logging! Expected analytics to contain identify , function(s).');
   });
 });
```
Comment (ghost) [inline src/__tests__/index.js]: The only other test ideas I have would be if we can setup mock functions in the `analytics` object and the execute the `*WhenReady()` functions and see if those mocked functions were called.

**Review by ghost — APPROVED** [review]
I think this looks good. I need to check it out and test it when I have a minute. I don't recall the settings issue I had, so if this is working...maybe the issue was in my brain.

File: README.md (hunk 5 above)
Comment (cfnelson) [reply README.md:58]: @ccuilla Thanks for spotting, have fixed the `changelog`.

File: examples/react-router/package.json (hunk 6 above)
Comment (cfnelson) [reply examples/react-router/package.json]: @ccuilla Yes, I left it out though as we `.gitignore` the `node_modules` if you try to `meteor npm install` it will error out. Will bump it to `^2.0.0` after we get the all clear.

Temp instructions to get the Meteor app running.
To get the example running I have been running `meteor npm install`, and copying over the `dist/index.umd.js` into `imports` and then replacing line 9 in`main.jsx` with `import { initAnalytics } from '../imports/index.umd';`.  Note: you need to first have run `npm install` in the root to trigger the build for the package.

File: src/__tests__/index.js (hunk 8 above)
Comment (cfnelson) [reply src/__tests__/index.js:22]: @ccuilla I like the idea! Updated to also check the `type`. 🔥

**Comment by Prtfw** [comment]
followed the readme, built minified files, router and custom event were logging. LGTM.

**Review by Prtfw — APPROVED** [review]

File: .babelrc [hunk 11]
```diff
@@ -1,3 +1,22 @@
 {
-  "presets": ["es2015", "stage-2"]
-}
\ No newline at end of file
+  "env": {
+    "build": {
+      "presets": [
+        [
+          "es2015",
```
Comment (ghost) [inline .babelrc:6]: Wonder if we should be using `babel-preset-env` now?

File: scripts/prepublish.sh [hunk 12]
```diff
@@ -1,5 +0,0 @@
-echo "> Start transpiling ES2015"
```
Comment (ghost) [inline scripts/prepublish.sh:1]: We're not doing this any more? Oh wait. I see it now...this is being done with rollup?

File: examples/react-router/client/main.jsx [hunk 13]
```diff
@@ -5,18 +5,28 @@ import { Meteor } from 'meteor/meteor';
...
                   && Meteor.settings.public
                   && Meteor.settings.public.analyticsSettings)
                   || false;
 
-// Initialize the @okgrow/auto-analytics package
-OKGAnalytics(SETTINGS);
+// NOTE: Initialize the @okgrow/auto-analytics package.
+initAnalytics({ analytics, integrations, options: {}, autorun: true });
+
+const exampleTrackEvent = () => analytics.track('Bought a Ticket', {
```
Comment (ghost) [inline examples/react-router/client/main.jsx:26]: So this is going directly to the analytics package in this case and not ours. Which is cool. Just making sure I understand.

File: scripts/prepublish.sh (hunk 12 above)
Comment (cfnelson) [reply scripts/prepublish.sh:1]: @ccuilla Yup, Rollup.js 👍  Gives us more control & can create multiple target builds, another big reason is we get a better bundle size. Current bundle/rollup setup can be further improved as well.

**Comment by ghost** [comment]
@cfnelson I think this is looking good. I have not yet pulled and tested it though.

File: examples/react-router/client/main.jsx (hunk 13 above)
Comment (cfnelson) [reply examples/react-router/client/main.jsx:26]: I guess maybe it should be or could be `window.analytics` as we assign it inside `intitAnalytics` to the window. I don't think the behaviour has changed in this PR. We always called the `analytics` object, its just we never had an example of `.track()` here. 

The main change this PR has is that we don't provide a default `analytics` bundle. The user has to go build it themselves. Created a bundle for the example to demonstrate the app :)

File: .babelrc (hunk 11 above)
Comment (cfnelson) [reply .babelrc:6]: Yes, we could switch this out if we liked. I think I may have attempted to use it at the time as well but it caused some odd builds which bugged out during testing.

However a lot has changed since this was originally written and we should make the change. I would like to add that I wish we didn't have to support `IE11`. 😹
